File: skills/businesses/lib/loader.py
#!/usr/bin/env python3
"""
businesses skill · loader

Carrega uma business inteira (manifest + employees + org-chart + routing) e
roda validação cruzada. Reusa os validators centralizados em
~/.claude/skills/_shared/validators/validators.py.

Use:
    from lib.loader import load_business, ValidationError
    biz = load_business('~/businesses/my-startup')
    print(biz.manifest.name, len(biz.employees))

ou via CLI:
    python3 lib/loader.py ~/businesses/my-startup
"""
from __future__ import annotations

import os
import re
import sys
import yaml
from dataclasses import dataclass
from pathlib import Path

# Importa validators centralizados
SHARED_VALIDATORS = os.path.expanduser('~/.claude/skills/_shared/validators')
if SHARED_VALIDATORS not in sys.path:
    sys.path.insert(0, SHARED_VALIDATORS)

from validators import (  # type: ignore[import-not-found]
    BusinessManifest,
    EmployeeFrontmatter,
    OrgChart,
    Routing,
    BusinessLoadContext,
    validate_business_integrity,
)


class ValidationError(Exception):
    """Raised quando uma business falha validação."""

    def __init__(self, message: str, errors: list[str] | None = None) -> None:
        super().__init__(message)
        self.errors = errors or []


@dataclass
class LoadedBusiness:
    """Container de business carregada e validada."""

    path: Path
    manifest: BusinessManifest
    employees: list[EmployeeFrontmatter]
    org_chart: OrgChart
    routing: Routing | None
    permanent_memory_path: Path | None


def _expand(p: str | Path) -> Path:
    return Path(os.path.expandvars(os.path.expanduser(str(p)))).resolve()


def _read_frontmatter_md(path: Path) -> tuple[dict, str]:
    """Lê arquivo .md, retorna (frontmatter dict, body)."""
    raw = path.read_text(encoding='utf-8')
    m = re.match(r'^---\n(.*?)\n---\n(.*)$', raw, flags=re.DOTALL)
    if not m:
        raise ValidationError(f'Frontmatter ausente ou malformado em {path}')
    fm_yaml = m.group(1)
    body = m.group(2)
    fm = yaml.safe_load(fm_yaml)
    if not isinstance(fm, dict):
        raise ValidationError(f'Frontmatter de {path} deve ser mapping')
    return fm, body


def load_business(path: str | Path, *, strict: bool = True) -> LoadedBusiness:
    """Carrega business completa de um diretório.

    Estrutura esperada:
        <path>/business.yaml
        <path>/employees/*.md
        <path>/org-chart.yaml
        <path>/routing.yaml         (opcional)
        <path>/escalation-triggers.yaml (opcional)
        <path>/memory/permanent.md  (opcional)

    Quando strict=True (default), lança ValidationError se algo está inválido.
    Quando strict=False, retorna a business com erros acumulados (raise apenas
    em erros fatais como manifest ausente).
    """
    biz_path = _expand(path)
    if not biz_path.is_dir():
        raise ValidationError(f'Diretório não encontrado: {biz_path}')

    errors: list[str] = []

    # 1. Manifest (obrigatório)
    manifest_path = biz_path / 'business.yaml'
    if not manifest_path.is_file():
        raise ValidationError(f'business.yaml ausente em {biz_path}')

    manifest_data = yaml.safe_load(manifest_path.read_text(encoding='utf-8'))
    try:
        manifest = BusinessManifest.model_validate(manifest_data)
    except Exception as exc:
        raise ValidationError(f'business.yaml inválido: {exc}') from exc

    # 2. Org chart (obrigatório)
    chart_path = biz_path / 'org-chart.yaml'
    if not chart_path.is_file():
        raise ValidationError(f'org-chart.yaml ausente em {biz_path}')

    chart_data = yaml.safe_load(chart_path.read_text(encoding='utf-8'))
    try:
        org_chart = OrgChart.model_validate(chart_data)
    except Exception as exc:
        raise ValidationError(f'org-chart.yaml inválido: {exc}') from exc

    # 3. Employees (obrigatório, ≥1)
    employees_dir = biz_path / 'employees'
    if not employees_dir.is_dir():
        raise ValidationError(f'employees/ ausente em {biz_path}')

    employees: list[EmployeeFrontmatter] = []
    for emp_file in sorted(employees_dir.glob('*.md')):
        try:
            fm, _body = _read_frontmatter_md(emp_file)
            employees.append(EmployeeFrontmatter.model_validate(fm))
        except (ValidationError, Exception) as exc:
            err = f'employee {emp_file.name}: {exc}'
            if strict:
                raise ValidationError(err) from exc
            errors.append(err)

    if not employees and strict:
        raise ValidationError(f'employees/ vazio em {biz_path}')

    # 4. Routing (opcional, documentação)
    # routing.yaml não é a fonte de verdade do roteamento em runtime: o router
    # consome auto_routes via registry._read_routing (que faz parse tolerante de
    # business.yaml + routing.yaml). Logo, um routing.yaml que não bate com o
    # schema canônico (formatos mais ricos: routing_rules, approval_gates, etc.)
    # NÃO deve invalidar uma business correta. Mantemos como warning.
    routing: Routing | None = None
    routing_path = biz_path / 'routing.yaml'
    if routing_path.is_file():
        routing_data = yaml.safe_load(routing_path.read_text(encoding='utf-8'))
        try:
            routing = Routing.model_validate(routing_data)
        except Exception:
            routing = None

    # 5. Permanent memory path (opcional)
    permanent_memory_path: Path | None = None
    permanent_md = biz_path / 'memory' / 'permanent.md'
    if permanent_md.is_file():
        permanent_memory_path = permanent_md

    # 6. Cross-protocol integrity check (BP7, intake único, sem ciclos, etc.)
    ctx = BusinessLoadContext(manifest=manifest, employees=employees, org_chart=org_chart)
    result = validate_business_integrity(ctx)
    if not result.valid:
        if strict:
            raise ValidationError(
                f'Integrity check falhou em {biz_path}', errors=result.errors
            )
        errors.extend(result.errors)

    if errors and strict:
        raise ValidationError(f'Business {manifest.name} tem erros', errors=errors)

    return LoadedBusiness(
        path=biz_path,
        manifest=manifest,
        employees=employees,
        org_chart=org_chart,
        routing=routing,
        permanent_memory_path=permanent_memory_path,
    )


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        print('Usage: python3 loader.py <business-path>', file=sys.stderr)
        return 2

    path = argv[1]
    try:
        biz = load_business(path)
    except ValidationError as exc:
        print(f'INVALID: {exc}', file=sys.stderr)
        for err in exc.errors:
            print(f'  - {err}', file=sys.stderr)
        return 1

    print(f'OK: {biz.manifest.name} v{biz.manifest.version}')
    print(f'  protocol: {biz.manifest.protocol}')
    print(f'  domains: {biz.manifest.domains}')
    print(f'  employees: {len(biz.employees)}')
    intake = next((e for e in biz.employees if e.is_brief_intake), None)
    print(f'  brief_intake: {intake.name if intake else "<NONE>"}')
    antagonists = [e.name for e in biz.employees if e.is_antagonist]
    print(f'  antagonists: {antagonists or "<none>"}')
    org_chart_nodes = len(biz.org_chart.chart) if (biz.org_chart and biz.org_chart.chart) else 0
    print(f'  org_chart nodes: {org_chart_nodes}')
    print(f'  routing: {"present" if biz.routing else "absent"}')
    print(f'  permanent_memory: {biz.permanent_memory_path or "<none>"}')
    return 0


if __name__ == '__main__':
    sys.exit(main(sys.argv))

File: skills/businesses/lib/loader.ts
#!/usr/bin/env bun
/**
 * businesses skill · loader (Bun/TypeScript)
 *
 * Loads a whole business (manifest + employees + org-chart + routing) and runs
 * cross-validation. Reuses the centralized validators in
 * ~/.nirvana/skills/_shared/validators/validators.ts.
 *
 * This is the Bun replacement for loader.py — Nirvana-OS runs on Bun alone, so
 * no Python is required on the host.
 *
 * Use:
 *   import { loadBusiness, ValidationError } from "./loader.ts";
 *   const biz = loadBusiness("~/businesses/my-startup");
 *
 * or via CLI (parity with the old loader.py output):
 *   bun loader.ts ~/businesses/my-startup
 */

import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import * as YAML from "yaml";
import { z } from "zod";

// Match python's yaml.safe_load leniency: duplicate keys don't throw (last
// wins). The default `yaml` parser is stricter and would reject manifests the
// canonical loader accepted.
const YAML_OPTS = { uniqueKeys: false } as const;
function parseYaml(text: string): unknown {
  return YAML.parse(text, YAML_OPTS);
}
import {
  BusinessManifestSchema,
  EmployeeFrontmatterSchema,
  OrgChartSchema,
  RoutingSchema,
  validateBusinessIntegrity,
  type BusinessLoadContext,
} from "../../_shared/validators/validators.ts";

export type BusinessManifest = z.infer<typeof BusinessManifestSchema>;
export type EmployeeFrontmatter = z.infer<typeof EmployeeFrontmatterSchema>;
export type OrgChart = z.infer<typeof OrgChartSchema>;
export type Routing = z.infer<typeof RoutingSchema>;

export class ValidationError extends Error {
  errors: string[];
  constructor(message: string, errors: string[] = []) {
    super(message);
    this.name = "ValidationError";
    this.errors = errors;
  }
}

export interface LoadedBusiness {
  path: string;
  manifest: BusinessManifest;
  employees: EmployeeFrontmatter[];
  org_chart: OrgChart;
  routing: Routing | null;
  permanent_memory_path: string | null;
}

function expand(p: string): string {
  let out = p;
  if (out.startsWith("~")) out = path.join(os.homedir(), out.slice(1));
  out = out.replace(/\$([A-Za-z_][A-Za-z0-9_]*)/g, (_, n) => process.env[n] ?? "");
  return path.resolve(out);
}

function formatZodError(err: z.ZodError): string {
  return err.issues
    .map((i) => `${i.path.join(".") || "(root)"}: ${i.message}`)
    .join("; ");
}

function readFrontmatterMd(file: string): { fm: Record<string, unknown>; body: string } {
  const raw = fs.readFileSync(file, "utf8").replace(/\r\n/g, "\n");
  const m = /^---\n([\s\S]*?)\n---\n([\s\S]*)$/.exec(raw);
  if (!m) throw new ValidationError(`Frontmatter ausente ou malformado em ${file}`);
  const fm = parseYaml(m[1]);
  if (fm === null || typeof fm !== "object" || Array.isArray(fm)) {
    throw new ValidationError(`Frontmatter de ${file} deve ser mapping`);
  }
  return { fm: fm as Record<string, unknown>, body: m[2] };
}

export function loadBusiness(inputPath: string, opts: { strict?: boolean } = {}): LoadedBusiness {
  const strict = opts.strict ?? true;
  const bizPath = expand(inputPath);
  if (!fs.existsSync(bizPath) || !fs.statSync(bizPath).isDirectory()) {
    throw new ValidationError(`Diretório não encontrado: ${bizPath}`);
  }

  const errors: string[] = [];

  // 1. Manifest (required)
  const manifestPath = path.join(bizPath, "business.yaml");
  if (!fs.existsSync(manifestPath)) throw new ValidationError(`business.yaml ausente em ${bizPath}`);
  const manifestParsed = BusinessManifestSchema.safeParse(parseYaml(fs.readFileSync(manifestPath, "utf8")));
  if (!manifestParsed.success) throw new ValidationError(`business.yaml inválido: ${formatZodError(manifestParsed.error)}`);
  const manifest = manifestParsed.data;

  // 2. Org chart (required)
  const chartPath = path.join(bizPath, "org-chart.yaml");
  if (!fs.existsSync(chartPath)) throw new ValidationError(`org-chart.yaml ausente em ${bizPath}`);
  const chartParsed = OrgChartSchema.safeParse(parseYaml(fs.readFileSync(chartPath, "utf8")));
  if (!chartParsed.success) throw new ValidationError(`org-chart.yaml inválido: ${formatZodError(chartParsed.error)}`);
  const orgChart = chartParsed.data;

  // 3. Employees (required, >= 1)
  const employeesDir = path.join(bizPath, "employees");
  if (!fs.existsSync(employeesDir) || !fs.statSync(employeesDir).isDirectory()) {
    throw new ValidationError(`employees/ ausente em ${bizPath}`);
  }
  const employees: EmployeeFrontmatter[] = [];
  const empFiles = fs.readdirSync(employeesDir).filter((f) => f.endsWith(".md")).sort();
  for (const empFile of empFiles) {
    try {
      const { fm } = readFrontmatterMd(path.join(employeesDir, empFile));
      const parsed = EmployeeFrontmatterSchema.safeParse(fm);
      if (!parsed.success) throw new ValidationError(formatZodError(parsed.error));
      employees.push(parsed.data);
    } catch (exc) {
      const err = `employee ${empFile}: ${(exc as Error).message}`;
      if (strict) throw new ValidationError(err);
      errors.push(err);
    }
  }
  if (employees.length === 0 && strict) throw new ValidationError(`employees/ vazio em ${bizPath}`);

  // 4. Routing (optional, documentation — tolerant parse, never invalidates)
  let routing: Routing | null = null;
  const routingPath = path.join(bizPath, "routing.yaml");
  if (fs.existsSync(routingPath)) {
    const parsed = RoutingSchema.safeParse(parseYaml(fs.readFileSync(routingPath, "utf8")));
    routing = parsed.success ? parsed.data : null;
  }

  // 5. Permanent memory path (optional)
  let permanentMemoryPath: string | null = null;
  const permanentMd = path.join(bizPath, "memory", "permanent.md");
  if (fs.existsSync(permanentMd)) permanentMemoryPath = permanentMd;

  // 6. Cross-protocol integrity check (BP7, single intake, no cycles, etc.)
  const ctx: BusinessLoadContext = { manifest, employees, org_chart: orgChart };
  const result = validateBusinessIntegrity(ctx);
  if (!result.valid) {
    if (strict) throw new ValidationError(`Integrity check falhou em ${bizPath}`, result.errors);
    errors.push(...result.errors);
  }

  if (errors.length > 0 && strict) throw new ValidationError(`Business ${manifest.name} tem erros`, errors);

  return {
    path: bizPath,
    manifest,
    employees,
    org_chart: orgChart,
    routing,
    permanent_memory_path: permanentMemoryPath,
  };
}

function main(argv: string[]): number {
  // Optional `--field <intake_employee|name|employees|all>` for programmatic use
  // (replaces the old loader-cli.py).
  const fieldIdx = argv.indexOf("--field");
  const field = fieldIdx !== -1 ? argv[fieldIdx + 1] : null;
  const pathArg = argv.find((a) => !a.startsWith("--") && a !== field);
  if (!pathArg) {
    process.stderr.write("Usage: bun loader.ts <business-path> [--field intake_employee|name|employees|all]\n");
    return 2;
  }
  let biz: LoadedBusiness;
  try {
    biz = loadBusiness(pathArg);
  } catch (exc) {
    if (exc instanceof ValidationError) {
      if (field) {
        process.stderr.write(`load_business failed: ${exc.message}\n`);
        return 1;
      }
      process.stderr.write(`INVALID: ${exc.message}\n`);
      for (const err of exc.errors) process.stderr.write(`  - ${err}\n`);
      return 1;
    }
    throw exc;
  }

  if (field) {
    const emps = biz.employees.map((e) => ({ name: e.name, is_brief_intake: !!e.is_brief_intake }));
    if (field === "intake_employee") {
      const i = biz.employees.find((e) => e.is_brief_intake);
      console.log(i ? i.name : "");
    } else if (field === "name") {
      console.log(biz.manifest.name);
    } else if (field === "employees") {
      console.log(JSON.stringify(emps));
    } else {
      console.log(JSON.stringify({ name: biz.manifest.name, employees: emps }, null, 2));
    }
    return 0;
  }

  const intake = biz.employees.find((e) => e.is_brief_intake);
  const antagonists = biz.employees.filter((e) => e.is_antagonist).map((e) => e.name);
  const orgChartNodes = biz.org_chart?.chart ? biz.org_chart.chart.length : 0;
  console.log(`OK: ${biz.manifest.name} v${biz.manifest.version}`);
  console.log(`  protocol: ${biz.manifest.protocol}`);
  console.log(`  domains: ${JSON.stringify(biz.manifest.domains)}`);
  console.log(`  employees: ${biz.employees.length}`);
  console.log(`  brief_intake: ${intake ? intake.name : "<NONE>"}`);
  console.log(`  antagonists: ${antagonists.length ? JSON.stringify(antagonists) : "<none>"}`);
  console.log(`  org_chart nodes: ${orgChartNodes}`);
  console.log(`  routing: ${biz.routing ? "present" : "absent"}`);
  console.log(`  permanent_memory: ${biz.permanent_memory_path || "<none>"}`);
  return 0;
}

if (import.meta.main) {
  process.exit(main(process.argv.slice(2)));
}
